Add full_token and tap entries to cask json output

cdalvaro opened the pull request:
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?
- [x] Have you successfully run `brew man` locally and committed any changes?

-----

This PR closes #10539

Rylan12 commented:
Thanks, @cdalvaro! I think this makes sense. We already have `full_name` in the formula json output, so adding `full_token` here for casks too is a good idea.

I think it would make sense to add a `tap` entry in the json output as well (for both formulae and casks).

I'm thinking something like this:

```json
{
  "formulae": [
    {
      "name": "foo",
      "full_name": "foo",
      "tap": "homebrew/core",
      ...
    }
  ],
  "casks": [
    {
      "token": "foo",
      "full_token": "homebrew/cask/foo",
      "tap": "homebrew/cask",
      ...
    }
  ]
}
```

Would you be willing to implement this?

cdalvaro commented:
Perfect!

I have already added the `full_token` key in #10539

I will block temporary the PR until I have added the `tap` key for both `formulae` and `casks`.

Rylan12 commented:
Awesome! Thanks, @cdalvaro!

Is there a reason this should still be a draft?

cdalvaro commented:
I was looking for the right way to add a test for the `tap` entry for formulae.

I think that test: `test/cmd/list_spec.rb` is the right place to add it, but this is my first contribution to the _brew_ project and I'm not very used to ruby 😅

ehawthorne821 commented:
What can I do for you? I’m kinda unavailable right these few weeks. Can I take a rain check on that ?

Mr. Earl GrayHawthorne, Owner 
Hawthorne-s Swirl Team, LLC
Commercial / Residential 
(H-VAC)Home Improvement  and Repairs Services.
410 Oneida Pl. NW 
WDC 20011-2151
Phone:
+1(202) 276-0184
Fax:
+1(443) 352-8122
Email:
earlgrayhawthorne@gmail.con
earlhawthorne70@gmail.com
Google:
@hawthorne-s-swirl-team
https://g.page/hawthorne-s-swirl-team 


> On Feb 6, 2021, at 9:41 PM, skysky365 <notifications@github.com> wrote:
> 
> ﻿
> @skysky365 approved this pull request.
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.

Rylan12 commented on Library/Homebrew/cask/cask.rb:
```diff
@@ -169,6 +169,8 @@ def eql?(other)
     def to_h
       {
         "token"          => token,
+        "full_token"     => "#{tap.name}/#{token}",
```
```suggestion
        "full_token"     => full_name,
```

We have a `full_name` method

Rylan12 replied:
I don't know if `full_name` or `full_token` is better to use here. Using token seems more consistent with the token/name difference between formulae and casks, but the method is called `full_name`.

Maybe someone from @Homebrew/cask can weigh in on what's best for consistency here.

cdalvaro replied:
OK! I'll replace it by `full_name`. I don't mind change it later to `full_token` if it is better that way.

cdalvaro replied:
One question here.

`full_name` is always `tap.name/token`? Or it is replaced by `token` when `tap.name` value is `homebrew/cask`?

I have seen that `full_name` for formulae is the same as `name` when the tap is `homebrew/core`

cdalvaro replied:
Done!

cdalvaro replied:
> One question here.
> 
> `full_name` is always `tap.name/token`? Or it is replaced by `token` when `tap.name` value is `homebrew/cask`?
> 
> I have seen that `full_name` for formulae is the same as `name` when the tap is `homebrew/core`

I've tested this change, and `full_name` is the same as `token` when taps are `homebrew/cask` and `homebrew/cask-fonts`

cdalvaro replied:
Tests `test/cask/cmd/list_spec.rb` have failed because of this issue.

Should I change the tests to take into account this behavior? Or do we always want to retrieve the tap name inside the `full_name`?

Rylan12 replied:
That's correct. `full_name` is identical to `name`/`token` for official taps (meaning taps that are `homebrew/*`). This is true for both formulae and casks.

> Should I change the tests to take into account this behavior?

Yes, please!

cdalvaro replied:
Done!

Rylan12 commented on Library/Homebrew/test/cmd/info_spec.rb:
```diff
@@ -26,6 +39,13 @@
       .and be_a_success
   end
 
+  it "check --json=v2 format", :integration_test do
+    setup_test_formula "testball"
+
+    expect { brew "info", "testball", "--json=v2" }
+      .to output(expected_output).to_stdout
+  end
+
```
Let's not add this as integration tests are super slow. Instead, how about using the tests in [`formula_spec.rb`](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/test/formula_spec.rb):

https://github.com/Homebrew/brew/blob/f416f8b54cec576a6528447f16b9b9bdd4df4b7d/Library/Homebrew/test/formula_spec.rb#L826-L843

cdalvaro replied:
Done! I have removed the original test.

Rylan12 commented on Library/Homebrew/cask/cask.rb:
```diff
@@ -169,6 +169,8 @@ def eql?(other)
       {
         "token"          => token,
+        "full_token"     => "#{tap.name}/#{token}",
+        "tap"            => tap.name,
```
```suggestion
        "tap"            => tap&.name,
```

`tap` can be `nil` so let's ensure there are no errors thrown in those cases

cdalvaro replied:
Perfect! I'll make the change!

cdalvaro replied:
Done!

Rylan12 commented on Library/Homebrew/formula.rb:
```diff
@@ -1766,6 +1766,7 @@ def to_hash
     hsh = {
       "name"                     => name,
       "full_name"                => full_name,
+      "tap"                      => tap.name,
```
```suggestion
      "tap"                      => tap&.name,
```

Same here: `tap` can be `nil` so we need to check for that.

cdalvaro replied:
Perfect! I'll make the change!

cdalvaro replied:
Done!

Rylan12 reviewed (commented):
Looking good! I few suggestions:

cdalvaro commented:
CI / Syntax check has failed installing `hadolint`

MikeMcQuaid commented:
@cdalvaro Please rebase this on `origin/master`; we had to change the CI configuration. Thanks!

cdalvaro commented:
Done!

Rylan12 commented:
Well done! Thank you so much for your first Hombrew PR, @cdalvaro! 🎉🥳

cdalvaro commented:
Thank you for your help, @Rylan12!

I'm very happy to have contributed to this great project! 🎉
